packages/ember-metal/lib/namespace_search.js

```diff
@@ -0,0 +1,200 @@
+import { getName, setName } from 'ember-utils';
+import { context } from 'ember-environment';
+
+// TODO, this only depends on context, otherwise it could be in utils
+// move into its own package
+// it is needed by Mixin for classToString
+// maybe move it into environment
+
+const hasOwnProperty = Object.prototype.hasOwnProperty;
+
+let searchDisabled = false;
+
+const flags = {
+  _set: 0,
+  _unprocessedNamespaces: false,
+  get unprocessedNamespaces() {
+    return this._unprocessedNamespaces;
+  },
+  set unprocessedNamespaces(v) {
+    this._set++;
+    this._unprocessedNamespaces = v;
+  }
+};
+
+let unprocessedMixins = false;
+
+export const NAMESPACES = [];
+export const NAMESPACES_BY_ID = Object.create(null);
+
+export function addNamespace(namespace) {
+  flags.unprocessedNamespaces = true;
+  NAMESPACES.push(namespace);
+}
+
+export function removeNamespace(namespace) {
+  let id = namespace.toString();
+  if (id) {
+    delete NAMESPACES_BY_ID[id];
+    if (namespace === context.lookup[id]) {
+      context.lookup[id] = undefined;
+    }
+  }
+
+  NAMESPACES.splice(NAMESPACES.indexOf(namespace), 1);
+}
+
+export function findNamespaces() {
+  if (!flags.unprocessedNamespaces) {
+    return;
+  }
+  let lookup = context.lookup;
+  let keys = Object.keys(lookup);
+  for (let i = 0; i < keys.length; i++) {
+    let key = keys[i];
+    // Only process entities that start with uppercase A-Z
+    if (!isUppercase(key.charCodeAt(0))) {
+      continue;
+    }
+    let obj = tryIsNamespace(lookup, key);
+    if (obj) {
+      setName(obj, key);
+    }
+  }
+}
+
+export function findNamespace(name) {
+  if (!searchDisabled) {
+    processAllNamespaces();
+  }
+  return NAMESPACES_BY_ID[name];
+}
+
+export function processNamespace(namespace) {
+  _processNamespace([namespace.toString()], namespace, new Set());
+}
+
+export function processAllNamespaces() {
+  let unprocessedNamespaces = flags.unprocessedNamespaces;
+  if (unprocessedNamespaces) {
+    findNamespaces();
+    flags.unprocessedNamespaces = false;
+  }
+
+  if (unprocessedNamespaces || unprocessedMixins) {
+    let namespaces = NAMESPACES;
+
+    for (let i = 0; i < namespaces.length; i++) {
+      processNamespace(namespaces[i]);
+    }
+
+    unprocessedMixins = false;
+  }
+}
+
+export function classToString() {
+  let name = getName(this);
+  if (name !== void 0) {
```

Comment:
even if it was `undefined` babel would replace it to `void 0`, no ?

Reply:
Yes, I believe so